genesis/utils/path_planing.py

```diff
@@ -0,0 +1,995 @@
+from abc import ABC, abstractmethod
+import taichi as ti
+import torch
+import torch.nn.functional as F
+import time
+import genesis as gs
+import genesis.utils.geom as gg
+
+
+class PathPlanner(ABC):
+    def __init__(self, entity):
+        self._entity = entity
+        self._solver = entity._solver
+
+        for joint in entity.joints:
+            if joint.type == gs.JOINT_TYPE.FREE:
+                gs.raise_exception("planning for the gs.JOINT_TYPE.FREE is not supported (yet)")
+            elif joint.type == gs.JOINT_TYPE.SPHERICAL:
+                gs.raise_exception("planning for the gs.JOINT_TYPE.SPHERICAL is not supported (yet)")
+
+    @abstractmethod
+    def plan(
+        self,
+        qpos_goal,
+        qpos_start=None,
+    ): ...
+
+    def get_link_transformation(self, robot_g_link_idx, obj_g_link_idx):
+        """
+        get transformation from robot link to the object link
+        robot_g_link_idx: int
+            global link idx of the link of the robot
+        obj_g_link_idx: int
+            global link idx of the base link of the object
+        """
+        robot_T = gg.trans_quat_to_T(
+            self._solver.get_links_pos(links_idx=robot_g_link_idx),
+            self._solver.get_links_quat(links_idx=robot_g_link_idx),
+        )
+        obj_T = gg.trans_quat_to_T(
+            self._solver.get_links_pos(links_idx=obj_g_link_idx), self._solver.get_links_quat(links_idx=obj_g_link_idx)
+        )
+        T_robot_object = torch.linalg.inv(robot_T) @ obj_T
+        return T_robot_object
+
+    def update_object(self, ee_link_idx, obj_link_idx, T_robot_object, envs_idx):
```

Comment:
I don't think storing pose in `T_robot_object` is a good idea. Just keep them as `_pos, _quat` pair. It would be more efficient and consistent with the codebase.